pkg/extensions/tbtc/recovery/derive_address_test.go

```diff
@@ -0,0 +1,183 @@
+		"ypub6Z7s8wJuKsxjd16oe85WH1uSbcbbCXuMFEhPMgcf7jQqNhQbT9jE52XVu1eBe18q2J3LwnDd54ufL2jNvidjfCkbd34aVwLtYdztLUqucwR",
+		4,
+		"3FQZbk6HBX72j2yyT5L8hWTymaJgwvb9u9",
+	},
+
+	"BIP141: P2WPKH zpub from m/0": {
+		"zpub6nBt2B13YbbubMCzAmoBLg4emchqHqyMQ9yLfzUwsuJvYgUHAcgLXjfqqHrprso814Croc8rheqMhV4h796L9dF67qFjoWrFC8FnLsVHknB",
+		0,
+		"bc1q8dnmvgj4jsvafe0wuwdm89aua2405jp7jp2zhn",
+	},
+	"BIP141: P2WPKH zpub from m/6'": {
+		"zpub6nBt2B1BtG8t2jc146zg8M8pjbffVk9UMXP6XJpcK1iyzww8VjqMDnRybqzahD3RBMu5XVjxZPdkSikyxscnxxRpjr777yvnRRQRDVAG58G",
+		7,
+		"bc1q4nrgadrcxz4tcqke9eh7f6zn8lxg8lz49d8v6k",
+	},
+	"BIP141: P2WPKH zpub from m/6'/0": {
+		"zpub6p6mUAk2dpLVVsguhHA27Qgd8e4q394Csha9jfAJCABrFRcnSv6AYPsgmAzgR8feBR4Spu2piv4xMcneocZajEvKtoHp111pizpe6aAEqfp",
+		0,
+		"bc1q9wwwcgcl2lw74quetxan4j6vhluyvyhy3dwt5l",
+	},
+	"BIP141: P2WPKH zpub from m/6'/4'": {
+		"zpub6p6mUAkAyUsTr1mTe4s3WsRWpxbWUD4grfoE4wf4EscKt2gJ7LYaHQnqRAMk6jtiNAc9qTrATeo4DCXUtjpV3gHN8AWtfcntZhsP5UeStdx",
+		1,
+		"bc1qwsuszzk93puxlgs6l6f54r66dev882pm2760dv",
+	},
+	"BIP141: P2WPKH zpub from m/6'/4'/9'": {
+		"zpub6rXD5NEyWK1YKTrPzAK96Yri6Y7EnBXFj7wsVumD2JJCmiudVMtieG8MRpKnYjNANEWZKBSefxF1MvBcH7E57anyYBBj6ms2XZMVgsuPzzs",
+		11,
+		"bc1qsszrcep8whzqh93ksmmckn77eh9fl55s5dzjx6",
+	},
+	"BIP141: P2WPKH zpub at m/6'/4'/9'/0/4": {
+		"zpub6sx8SbypUZWDUJHvUUs8V6zwmak399trAMDc95WYVjniRoDphotnh6BdvDbmdunkRwA9hFpBXjGDDKLweR3kTSSCVNm15rANpN4XixewDwG",
+		4,
+		"bc1q5l3j7e2s3dzg4vxquxldvzw4dwdlvvhdc9c4zh",
+	},
+}
+
+func TestDeriveAddress(t *testing.T) {
+	for testName, testData := range deriveAddressTestData {
+		t.Run(testName, func(t *testing.T) {
+			address, err := DeriveAddress(testData.extendedAddress, uint32(testData.addressIndex))
+
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			if address != testData.expectedAddress {
+				t.Errorf(
+					"unexpected derived address\nexpected: %s\nactual:   %s",
+					testData.expectedAddress,
+					address,
+				)
+			}
+		})
+	}
+}
+
+var deriveAddressTestFailureData = map[string]struct {
+	extendedAddress string
+	addressIndex    int
+	failure         string
+}{
+	"BIP141 P2WPKH nested in P2SH ypub at m/6'/4'/9'/0/11'": {
+		"ypub6Z7s8wJuKsxjd16oe85WH1uSbcbbCXuMFEhPMgcf7jQqNhQbT9jE52XVu1eBe18q2J3LwnDd54ufL2jNvidjfCkbd34aVwLtYdztLUqucwR",
+		11 + 2147483648,
+		"cannot derive a hardened key from a public key",
+	},
+
+	"BIP141 P2WPKH nested in P2SH ypub at m/6'/4'/9'/0/11' with a private key": {
+		"yprvAL8WjRn1VWQSQX2LY6YVusxi3am6o5BVt1mnZJD3ZPsrVu5SucQyXED23ikCvDeeFHTMeX9q5n5MHNTLWQvCSm3KWnA3KdyZuDXncTn2VW5",
+		11 + 2147483648,
+		"unsupported public key format",
+	},
+
+	"BIP141 ypub is too deep at m/0/0/0/0/0/0": {
+		"ypub6bp11ZqNVMqm3C3eXAFGpEvKqNfEZ6Vhznd4Uo3S73RYTSFgmF7q9sWPoCFhLGVMSLqKZZpcpHoKgHNwStDuqQPnDfF13goQwS8qSFA6vnz",
+		0,
+		"extended public key is deeper than 4, depth: 5",
+	},
+}
+
+func ErrorContains(err error, expected string) bool {
+	return strings.Contains(err.Error(), expected)
+}
```

Comment:
Incidentally I'm seeing just now that errors got some boosts in Go 1.13 we haven't really been making use of anywhere: https://blog.golang.org/go1.13-errors cc @keep-network/go